### ISSUE TITLE
Provide `BufReader` and `BufWriter`

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio-io#21]
Similar to the `std` types, but these would impl `AsyncRead` and `AsyncWrite`, with optimized implementations for `read_buf` and `write_buf`.

**Comment by alexcrichton** [comment]
Oh I added impls of AsyncRead/Write for the std ones, but we should probably override the read_buf/write_buf methods.

**Comment by carllerche** [comment]
I tried providing specialized implementations of `read_buf` and `write_buf` against the std types, however given that the buffer is not exposed, it was not possible to get desired semantics.

Specifically, it wasn't possible to have read_buf and write_buf fill the buffer while also limiting it to at most one sys call.

Given this, I think the correct way to move forward is to provide our own versions of `BufReader` and `BufWriter` in `tokio-io`.

**Comment by alexcrichton** [comment]
I'd also be fine in parallel exploring solutions upstream as well (as this likely isn't *just* related to this lib's usage of those types)

**Comment by izderadicka** [comment]
Also tokio-io provides functions read_untils and lines which require type implementing AsyncRead + BufRead, however it looks like the implementation of such type it still missing in tokio. Or am I missing something?

**Comment by carllerche** [comment]
The tokio-io crate has been moved into the [tokio](https://github.com/tokio-rs/tokio) git repo.

This issue has been inactive, so I am not going to open a new one in the tokio repo. However, if anyone wishes to continue this discussion, please feel free to open a new issue over there.
